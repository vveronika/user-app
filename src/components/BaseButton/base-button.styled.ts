import styled, { ThemedStyledProps } from 'styled-components';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { theme } from 'app.styled';

interface ButtonProps {
  appearance?: 'primary' | 'action' | 'warning';
}

const getStyles = (
  props: ThemedStyledProps<ButtonProps, any>,
  property: string,
) => {
  if (props.appearance === 'primary') {
    return props.theme[property][props.appearance];
  } else if (props.appearance === 'action') {
    return props.theme[property][props.appearance];
  } else if (props.appearance === 'warning') {
    return props.theme[property][props.appearance];
  }
  return props.theme[property].default;
};

export const Button = styled.button<ButtonProps>`
  background-color: ${(props) => getStyles(props, 'background')};
  color: ${(props) => getStyles(props, 'font')};
  border: 1px solid ${(props) => getStyles(props, 'border')};
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
  height: 40px;
  border-radius: 5px;
  margin: 5px 0;
  outline: none;
  transition: ${theme.buttonTransition};

  &:hover {
    cursor: pointer;
    background-color: ${(props) => getStyles(props, 'backgroundHover')};
    color: ${(props) => getStyles(props, 'fontHover')};
    border: 1px solid ${(props) => getStyles(props, 'borderHover')};
  }

  &:disabled {
    background-color: ${theme.greenLight};
    color: ${theme.white};
    border: ${theme.greenLight};
  }
`;

export const Icon = styled(FontAwesomeIcon)`
  margin-right: 5px;
  height: 11px;
`;
