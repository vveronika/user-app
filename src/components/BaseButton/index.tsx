import React from 'react';
import { Button, Icon } from 'components/BaseButton/base-button.styled';
import { IconName } from '@fortawesome/fontawesome-svg-core';
import Loader from 'components/Loader';
import { theme } from 'app.styled';

interface Props {
  label: string;
  onClick: () => void;
  appearance?: 'action' | 'primary' | 'warning';
  disabled?: boolean;
  icon?: IconName;
  loading?: boolean;
}

const BaseButton: React.FC<React.PropsWithChildren<Props>> = (props) => {
  const { label, onClick, appearance, disabled, icon, children } = props;
  return (
    <Button disabled={disabled} appearance={appearance} onClick={onClick}>
      {!props.loading && icon && <Icon icon={icon} />}
      {!props.loading && children}
      {!props.loading && label}
      {props.loading && <Loader color={theme.white} size={20} />}
    </Button>
  );
};

export default BaseButton;
