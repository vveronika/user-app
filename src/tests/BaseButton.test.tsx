import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import BaseButton from 'components/BaseButton';

configure({ adapter: new Adapter() });

describe('Check if button with every appereance variant renders', () => {
  test('action button renders', () => {
    const wrapper = shallow(
      <BaseButton
        appearance="action"
        label="Action button"
        onClick={() => {}}
      />,
    );
    expect(wrapper.exists()).toBeTruthy;
  });
  test('warning button renders', () => {
    const wrapper = shallow(
      <BaseButton
        appearance="warning"
        label="Warning button"
        onClick={() => {}}
      />,
    );
    expect(wrapper.exists()).toBeTruthy;
  });
  test('primary button renders', () => {
    const wrapper = shallow(
      <BaseButton
        appearance="primary"
        label="Primary button"
        onClick={() => {}}
      />,
    );
    expect(wrapper.exists()).toBeTruthy;
  });
  test('default button renders', () => {
    const wrapper = shallow(
      <BaseButton label="Default button" onClick={() => {}} />,
    );
    expect(wrapper.exists()).toBeTruthy;
  });
});
