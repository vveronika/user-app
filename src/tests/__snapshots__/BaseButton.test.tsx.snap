// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check if button with every appereance variant renders action button renders 1`] = `ShallowWrapper {}`;

exports[`Check if button with every appereance variant renders default button renders 1`] = `ShallowWrapper {}`;

exports[`Check if button with every appereance variant renders primary button renders 1`] = `ShallowWrapper {}`;

exports[`Check if button with every appereance variant renders warning button renders 1`] = `ShallowWrapper {}`;
